File: test/rust/crates/specgate-fixtures/src/conformance/stateful/multi_mutation.rs
// Multiple mutations within a single operation.
// The system must capture every mutation of count, not just boundaries.
use specgate::*;

#[spec_setup("increment_twice")]
pub fn make_counter() -> Counter {
    Counter { count: 0 }
}

#[derive(SpecEvent)]
pub struct Counter {
    #[spec_event]
    pub count: i32,
}

impl Counter {
    #[spec_operation("increment_twice")]
    pub fn increment_twice(&mut self) {
        self.count += 1;
        self.count += 1;
    }
}

File: test/csharp/SpecGateFixtures/Conformance/Stateful/MultiMutation/multi_mutation.cs
using SpecGate.Annotations;
namespace SpecGateFixtures.Conformance.Stateful.MultiMutation;

/// <summary>
/// State machine whose operation mutates the same field twice, verifying that
/// every intermediate mutation is captured (not just the boundary values).
/// </summary>
public class Counter
{
    /// <summary>The current count; each assignment is captured as a distinct mutation.</summary>
    [SpecEvent("count")]
    public int Count { get; set; }

    /// <summary>Builds a counter starting at zero.</summary>
    /// <returns>A new <see cref="Counter"/> with <see cref="Count"/> 0.</returns>
    [SpecSetup("increment_twice", Spec = "fixture.multi_mutation")]
    public static Counter Make() => new() { Count = 0 };

    /// <summary>Increments the counter twice, emitting both intermediate values.</summary>
    [SpecOperation("increment_twice", Spec = "fixture.multi_mutation")]
    public void IncrementTwice()
    {
        Count += 1;
        Count += 1;
    }
}
